src/lib.rs

```diff
@@ -560,23 +387,109 @@ impl<'hedwig, T, P> HedwigPublishBuilder<'hedwig, T, P> {
                 schema_url,
                 FORMAT_VERSION_V1,
             )
-            .map_err(PublishError::MessageSerialization)?;
+            .map_err(Error::MessageSerialize)?;
         let route = (self.hedwig.message_router)(data_type, converted.format_version.0)
-            .ok_or_else(|| PublishError::MessageRouting(converted.id))?;
+            .ok_or_else(|| Error::MessageRoute(converted.id))?;
         self.messages.push((route, converted));
         Ok(self)
     }
 
-    /// Publish all the messages.
-    pub fn publish(self) -> Result<P::MessageIds, PublishError>
+    /// Publish all the messages
+    ///
+    /// Does not consume the builder. Will return `Ok` only if and when all of the messages from
+    /// the builder have been published successfully. In case of failure, unpublished messages will
+    /// remain enqueued in this builder for a subsequent publish call.
```

Comment:
ooh nice!

Reply:
actually thinking some more - does this mean you've to keep multiple batches around now? Since you'd queue a publish and then start building a new batch, but the old one might have failed and so on.. Should this instead have a single batch and then a "pending messages" that are moved into the batch on next publish call ?

Reply:
A good question.

You would indeed be able to either wait for `publish` future to complete or add to this batch, not both at the same time. I think for most use-cases this should be fine, and if users want an ability to both create the messages and publish at the same time they should introduce some sort of a buffer or channel in between the async task that’s creating and publishing batches and the other tasks that create messages to be published.